File: nqueen.py
#! /usr/local/bin/python3
class Solution:
    def nQueens(self, n):
        result = []
        cols, diag1, diag2 = set([]), set([]), set([])
        def helper(row, arr, n):
            for col in range(n):
                if col in cols:
                    continue
                if (row - col) in diag1:
                    continue
                if (row + col) in diag2:
                    continue
                if row == n - 1:
                    arr.append(col)
                    arr2 = arr.copy()
                    result.append(arr2)
                    arr.pop()
                else:
                    cols.add(col)
                    diag1.add(row - col)
                    diag2.add(row + col)
                    arr.append(col)
                    helper(row + 1, arr, n)
                    arr.pop()
                    cols.remove(col)
                    diag1.remove(row - col)
                    diag2.remove(row + col)

        helper(0, [], n)
        return result
    


n = 8
s = Solution()
result = s.nQueens(n)
for arr in result:
    print(arr)

for arr in result:
    for i in range(n):
        for j in range(n):
            if j == arr[i]:
                print('Q', end= '')
            else:
                print('.', end = '')
        print('')
    print()

# return the count of Nqueen solutions
def totalNQueens(self, n: int) -> int:
    cols, diag1, diag2 = set([]), set([]), set([])
    def helper(row, count, n):
        for col in range(n):
            if col in cols:
                continue
            if (row - col) in diag1:
                continue
            if (row + col) in diag2:
                continue
            if row == n - 1:
                count += 1
            else:
                cols.add(col)
                diag1.add(row - col)
                diag2.add(row + col)
                count = helper(row + 1, count, n)
                cols.remove(col)
                diag1.remove(row - col)
                diag2.remove(row + col)
        return count

    return helper(0, 0, n)

File: nqueen.js
var totalNQueens = function(n) {
    const cols = {};
    const diag1 = {};
    const diag2 = {};
    const results = [];

    function helper(row, count, n) {
        for (let col = 0; col < n; col++) {
            if (cols[col]) continue;
            const sum = row + col;
            const sub = row - col;
            if (diag1[sum]) continue;
            if (diag2[sub]) continue;

            if (row == n - 1) {
                count++;
            } else {
                cols[col] = true;
                diag1[sum] = true;
                diag2[sub] = true;
                count = helper(row + 1, count, n);
                delete cols[col];
                delete diag1[sum];
                delete diag2[sub];
            }
        }
        return count;
    }

    /**
       this function traces the result path.
    **/
    function helper2(row, trace, n) {
      for (let col = 0; col < n; col++) {
        if (cols[col]) continue;
        const sum = row + col;
        const sub = row - col;
        if (diag1[sum]) continue;
        if (diag2[sub]) continue;

        if (row == n - 1) {
          trace.push(col);
          results.push([...trace]);
          trace.pop();
        } else {
          cols[col] = true;
          diag1[sum] = true;
          diag2[sub] = true;
          trace.push(col);
          helper2(row + 1, trace, n);
          trace.pop(col);
          delete cols[col];
          delete diag1[sum];
          delete diag2[sub];
        }
      }
    }

    // return helper(0, 0, n);
    helper2(0, [], n);
    return results;
};

let n = 8;
let r = totalNQueens(n);

for(let s of r) {
  for(let row of s) {
    let rowDisplay = '';
    for (let col=0; col<n; col++) {
      const symbol = col === row ? 'Q':'_';
      rowDisplay += symbol;
    }
    console.log(rowDisplay);
  }
  console.log();
}
console.log(`result for [${n}*${n}] is ${r.length}.`);
